service/history/queue/domain_filter.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queue
+
+// NewDomainFilter creates a new domain filter
+func NewDomainFilter(
+	domainIDs map[string]struct{},
+	reverseMatch bool,
+) DomainFilter {
+	return DomainFilter{
+		DomainIDs:    domainIDs,
+		ReverseMatch: reverseMatch,
```

Comment:
I think I asked this before. But could you give me an example of this reverse match use case?

Reply:
This is basically for the default queue which is responsible for all the domains in the beginning. Instead of listing all domainIDs here and keep refreshing to see if there's new domains, I think using an empty list and do a reverse match is a much better approach. 

Later if a split is performed on the default queue and a domain no longer belongs to the default queue, reverse match will still be true and domainID list will contain the one that got split out. This means now the default queue is responsible for all domains except the one that has been split out.

Reply:
It is weird to me because the Filter/Include/Exclude have different behavior depends on the domain filter initialization. If it is reverse matche is true, then exclude is include and include is exclude.

Reply:
Synced offline. `Include` will always add domains to the set, `Exclude` will always remove domains from the set and `Merge` always output a union of two sets.